input/script_utils.go

```diff
@@ -1467,6 +1467,41 @@ func CommitScriptAnchor(key *btcec.PublicKey) ([]byte, error) {
 	return builder.Script()
 }
 
+// TaprootOutputKeyAnchor returns the segwit v1 (taproot) witness program that
+// encodes the anchor output spending conditions: the passed key can be used
+// for keyspend, with the OP_CSV 16 clause living within an internal tapscript
+// leaf.
+//
+// Spend paths:
+//   - Key spend: <key_signature>
+//   - Script spend: OP_16 CSV <control_block>
```

Comment:
It's the to_local and to_remote internal key. They have different scripts so the to_local uses the `local_delayedpubkey` and the to_remote uses the `remotepubkey`. So a 3rd party must wait for those scripts to be claimed

Reply:
Good question, we went thru a few iterations here so far to arrive at this point. Zooming out, the main goal is that 3rd parties can still spend after a period of time. As is, it's the actual top-level to_local+to_remote _internal_ keys. Both of these can only be spent via a script path, so only after the main outputs have been swept can a 3rd party sweep as Eugene mentions above. 

Agree we can add more rationale here in the spec as well. Reflecting a bit, I think I wanted to avoid a "double tweak" scenario here, where'd we'd take an existing key, then tweak it again. I wasn't sure if ppls libraries supported something like that (but pretty sure we do). So the question is do we go with that route to allow 3rd parties to sweep asap (say based on the multi-sig taproot output key), or keep as is?

Reply:
> As is, it's the actual top-level to_local+to_remote internal keys. Both of these can only be spent via a script path, so only after the main outputs have been swept can a 3rd party sweep 

But what if the `to_local` is spent via the revocation path? then that internal key is never revealed meaning a third party wont be able to figure out how to spend one of the anchor outputs. right? 

> So the question is do we go with that route to allow 3rd parties to sweep asap (say based on the multi-sig taproot output key),

I think if the multi-sig taproot output key was used then that would require another sign&nonce exchange between the peers though since they expect to use the keypath if they want to CPFP... right?

Any reason not to just use the `to_local` and `to_remote` *output* keys?

Reply:
> But what if the to_local is spent via the revocation path? then that internal key is never revealed meaning a third party wont be able to figure out how to spend one of the anchor outputs. right?

Correct. 

> I think if the multi-sig taproot output key was used then that would require another sign&nonce exchange between the peers though since they expect to use the keypath if they want to CPFP... right?

So this is what's used as the internal key. You can't make the control block without the internal key. If it's the multi-sig output key, then that works as well. 

> Any reason not to just use the to_local and to_remote output keys?

This works just like the multi-sig key. As mentioned above, initially I was trying to avoid a double tweak on a private key. It doesn't seem avoidable though ultimately, our libraries support it, but we'll need to make sure the popular libraries used by the others do as well.

Reply:
> So this is what's used as the internal key. You can't make the control block without the internal key. If it's the multi-sig output key, then that works as well.

But surely we want to make the keyspend path spendable by the party who's anchor output this is? ie, the keyspend path, and hence the internal key, should be something that they can sign for on their own.

Reply:
> But what if the to_local is spent via the revocation path? then that internal key is never revealed meaning a third party wont be able to figure out how to spend one of the anchor outputs. right?

This is a good point, we want to make sure the anchor is always spendable to prevent UTXO set bloat

> Any reason not to just use the to_local and to_remote output keys?

I'm not sure if I understand the suggestion here, but if this is the non-breach scenario, I don't think either party can sign for something encumbered by the `to_local` output key since it is a tweaked `revocation_pubkey` and the revocation secret will still be a secret

Reply:
> the revocation secret will still be a secret

Ah yes - my bad

Reply:
> Does this mean we need to have a third anchor?
anchor_internal_key = remotepubkey/local_delayedpubkey/to_local_output_key

Adding a third anchor would break CPFP carve-out. Whether that's ok is probably a longer discussion. If you want to get into the weeds, here is where it would break in bitcoin core (https://github.com/bitcoin/bitcoin/pull/15681/files#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R614)

> Or a bizarre thought, disable the key path in to_local_output so the internal key is always revealed via script spend? much more expensive tho.

Could you go into more detail? I'm assuming this would still keep the `revocationpubkey` claim path in some form

Reply:
> Could you go into more detail? I'm assuming this would still keep the revocationpubkey claim path in some form

Was thinking the left leaf be the delay path, the right leaf be the revocation path but we also put the `local_delayedpubkey` there, something like `<revocation_pubkey> OP_CHECKSIGVERIFY <local_delayedpubkey> OP_DROP`. So the internal key of the anchor output is always revealed.

Also thought about some alternatives,
- sign two local commit txes, one being the anchor output using `revocationpubkey` as internal key and the other use the delayed key, but this complicates things too much.
- use `OP_CHECKSIG` to do the funding so the local key is always revealed. don't like it as musig2 is not used.
- disable key spend in anchor, and use `IF local sig ELSE timeout path` in one leaf. Maybe feasible, kinda defeat the purpose of tapscript tho.

Reply:
> `<revocation_pubkey> OP_CHECKSIGVERIFY <local_delayedpubkey>`

needs to be `OP_CHECKSIG` since *VERIFY doesn't put anything on the stack. also as you said, makes it more expensive to sweep during breach

> disable key spend in anchor, and use IF local sig ELSE timeout path in one leaf. Maybe feasible, kinda defeat the purpose of tapscript tho.

yeah does defeat the purpose of tapscript

Reply:
> Was thinking the left leaf be the delay path, the right leaf be the revocation path but we also put the local_delayedpubkey there, something like <revocation_pubkey> OP_CHECKSIGVERIFY <local_delayedpubkey> OP_DROP. So the internal key of the anchor output is always revealed.

plus one for the idea of making keyspend path a NUMS point and then adding revocation path as a tapscript with a forced reveal of the `local_delayedpubkey` 🎖️

Reply:
New version pushed with this change: to local now uses a script path to handle the revocation case. Tradeoff here is ensuring UTXO set can always be constrained, but it makes force close sweeping more expensive (relative to the keyspend for revocation). I think it might still be on par with the old revocation path though, since that has all the other keys in line (if/else, etc).